core/src/main/java/com/linecorp/armeria/client/ClientThreadLocalState.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright 2019 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.client;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.function.Consumer;
+
+import javax.annotation.Nullable;
+
+import com.google.common.collect.ImmutableList;
+
+import com.linecorp.armeria.common.util.SafeCloseable;
+
+final class ClientThreadLocalState {
+
+    private static final ThreadLocal<ClientThreadLocalState> threadLocalState = new ThreadLocal<>();
```

Comment:
I guess it might be 50/50 whether this is called from ab event loop or not. Should we use `FastThreadLocal`?